web/packages/teleport/src/Navigation/RecentHistory.tsx

```diff
@@ -0,0 +1,298 @@
+/**
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React, { useState, useEffect, useRef } from 'react';
+import styled from 'styled-components';
+import { matchPath } from 'react-router';
+import { NavLink } from 'react-router-dom';
+import { Flex, Text, H4, P3, ButtonIcon } from 'design';
+import { Cross } from 'design/Icon';
+
+import { TeleportFeature } from 'teleport/types';
+import { useFeatures } from 'teleport/FeaturesContext';
+
+import { getSubsectionStyles } from './SideNavigation/Section';
+import { SidenavCategory } from './SideNavigation/categories';
+
+export type RecentHistoryItem = {
+  category?: SidenavCategory;
+  title: string;
+  route: string;
+  exact?: boolean;
+};
+
+type AnimatedItem = RecentHistoryItem & {
+  animationState: 'exiting' | 'entering' | '';
+};
+
+function getIconForRoute(
+  features: TeleportFeature[],
+  route: string
+): (props) => JSX.Element {
+  const feature = features.find(feature =>
+    matchPath(route, {
+      path: feature?.route?.path,
+      exact: false,
+    })
+  );
+
+  const icon = feature?.navigationItem?.icon || feature?.topMenuItem?.icon;
+  if (!icon) {
+    return () => null;
+  }
+
+  return icon;
+}
+
+export function RecentHistory({
+  recentHistoryItems,
+  onRemoveItem,
+}: {
+  recentHistoryItems: RecentHistoryItem[];
+  onRemoveItem: (route: string) => void;
+}) {
+  const features = useFeatures();
+  const [animatedItems, setAnimatedItems] = useState<AnimatedItem[]>([]);
+  const prevItemsRef = useRef<RecentHistoryItem[]>([]);
+
+  useEffect(() => {
```

Comment:
nit: There is too much going on inside this useEffect. Looks like you can abstract most parts to a pure js function and use that in useEffect to update final state, will be cleaner that way. 

Feel free to do that later when you encounter this part of codebase again ;)